digdag-core/src/main/java/io/digdag/core/database/BasicDatabaseStoreManager.java

```diff
@@ -274,42 +283,105 @@ public int getIndex()
+        if (nested != null) {
+            return action.call(nested, nested.attach(daoIface));
         }
-        catch (InterruptedException ex) {
-            throw Throwables.propagate(ex);
+        else {
+            try {
+                return transactionRetryExecutor.runInterruptible(() -> {
+                    try (Handle handle = dbi.open()) {
+                        currentTransaction.set(handle);
+                        handle.begin();
+                        T retval;
+                        try {
+                            retval = action.call(handle, handle.attach(daoIface));
+                        }
+                        catch (Exception ex) {
+                            try {
+                                handle.rollback();
+                            }
+                            catch (Exception rollback) {
+                                ex.addSuppressed(rollback);
+                            }
+                            Throwables.propagateIfInstanceOf(ex, exClass1);
+                            Throwables.propagateIfInstanceOf(ex, exClass2);
+                            Throwables.propagateIfPossible(ex);
+                            throw new TransactionFailedException(
+                                    "Transaction failed due to exception being thrown " +
+                                    "from within the callback. See cause " +
+                                    "for the original exception.", ex);
+                        }
+                        handle.commit();
+                        return retval;
+                    }
+                    finally {
+                        currentTransaction.set(null);
+                    }
+                });
+            }
+            catch (RetryGiveupException ex) {
+                Throwable innerException = ex.getCause();
+                Throwables.propagateIfInstanceOf(innerException, exClass1);
+                Throwables.propagateIfInstanceOf(innerException, exClass2);
+                throw Throwables.propagate(innerException);
+            }
+            catch (InterruptedException ex) {
+                throw Throwables.propagate(ex);
+            }
         }
     }
 
     @SuppressWarnings("unchecked")
-    public <T, E1 extends Exception, E2 extends Exception> T transaction(TransactionActionWithExceptions<T, D, E1, E2> action, Class<E1> exClass1, Class<E2> exClass2) throws E1, E2
+    public <T, E1 extends Exception, E2 extends Exception> T transactionWithLocalRetry(RetryableTransactionActionWithExceptions<T, D, E1, E2> action, Class<E1> exClass1, Class<E2> exClass2) throws E1, E2
     {
+        // TODO
+        // Here should throw error if currentTransaction.get() != null to avoid race condition under high load.
+        // In following scenario, a deadlock happens:
+        // * thread1 acquires a row lock on RDBMS (such as lockTaskIfExists), then calls transactionWithLocalRetry.
```

Comment:
These logics of locking sounds a bit complicated to me. It may prevent other developers from trying to look into around here. Maybe we'd better simplify these lock logics or write the policies of lock logics as a document later.